proto/osmosis/streamswap/v1/tx.proto

```diff
@@ -0,0 +1,138 @@
+  // tokens_out from the pool and remained tokens_in. Returns error if it's
+  // called before the sale end.
+  rpc ExitSale(MsgExitSale) returns (MsgExitSaleResponse);
+
+  // FinalizeSale clean ups the sale and sends income (earned tokens_in) to the
+  // Sale recipient. Returns error if called before the Sale end. Anyone can
+  // call this method.
+  rpc FinalizeSale(MsgFinalizeSale) returns (MsgFinalizeSaleResponse);
+}
+
+message MsgCreateSale {
+  // Sale creator and the account which provides token (token_out) to the sale.
+  // When processing this message, token_out
+  string creator = 1;
+  // token_in is a denom used to buy `token_out`. May be referred as a
+  // "quote currency".
+  string token_in = 2;
+  // token_out is a coin supply (denom + amount) to sell. May be referred as
+  // "base currency". The whole supply will be transferred from the creator
+  // to the module and will be sold during the sale.
+  cosmos.base.v1beta1.Coin token_out = 3 [ (gogoproto.nullable) = false ];
+
+  // Maximum fee the creator is going to pay for creating a sale. The creator
+  // will be charged params.SaleCreationFee. Transaction will fail if
+  // max_fee is smaller than params.SaleCreationFee. If empty, creator
+  // accepts doesn't accept any fee.
+  repeated cosmos.base.v1beta1.Coin max_fee = 4
+      [ (gogoproto.nullable) = false ];
+
+  // start time when the token sale starts.
+  google.protobuf.Timestamp start_time = 5
+      [ (gogoproto.nullable) = false, (gogoproto.stdtime) = true ];
+  // duration time that the sale takes place over
+  google.protobuf.Duration duration = 6
+      [ (gogoproto.nullable) = false, (gogoproto.stdduration) = true ];
+
+  // Recipient is the account which receives earned `token_in` from when the
+  // sale is finalized. If not defined (empty) the creator
+  // account will be used.
+  string recipient = 7;
+
+  // Name for the sale, max 40 characters, min 4. Required.
+  string name = 8;
+  // URL with sale and project details. Can be a link a link to IPFS,
+  // hackmd, project page, blog post... Max 120 characters. Must be
+  // valid agains Go url.ParseRequestURI. Required.
+  string url = 9;
+}
+
+message MsgCreateSaleResponse {
+  uint64 sale_id = 1 [ (gogoproto.moretags) = "yaml:\"sale_id\"" ];
+}
+
+message MsgSubscribe {
+  // sender is an account address adding a deposit
+  string sender = 1;
+  // sale_id is a valid ID of this module Sale
+  uint64 sale_id = 2 [ (gogoproto.moretags) = "yaml:\"sale_id\"" ];
+  // number of sale.token_in staked by a user.
+  string amount = 3 [
+    (gogoproto.customtype) = "github.com/cosmos/cosmos-sdk/types.Int",
+    (gogoproto.nullable) = false
+  ];
+}
+
+message MsgWithdraw {
+  // sender is an account address subscribed to the sale_id
+  string sender = 1;
+  uint64 sale_id = 2 [ (gogoproto.moretags) = "yaml:\"sale_id\"" ];
+  // amount of in_tokens to withdraw, must be at most the amount of not spent
```

Comment:
nit: since that's how we consistently refer to these in the rest of the spec
```suggestion
  // amount of tokens_in to withdraw must be at most the amount of not spent
```